Serialize `package` and SPI doc comments in swiftsourceinfo

hamishknight opened the pull request:
Requestify raw and brief comment computation, and allow `package` and SPI doc comments to be serialized in swiftsourceinfo. In the future we may want to emit a separate private swiftdoc for SPI doc comments, but for now we can just include them in the swiftsourceinfo to at least allow them to be picked up for local builds.

hamishknight commented:
@swift-ci please test

bnbarham reviewed (approved):
😍

bnbarham commented on test/SourceKit/CodeComplete/complete_docbrief_package.swift:
```diff
@@ -0,0 +1,50 @@
+
+// RUN: %empty-directory(%t)
+// RUN: %empty-directory(%t/Modules)
+// RUN: %{python} %utils/split_file.py -o %t %s
```
Mind using `%split-file` instead? See the other tests, but it uses `//--- <name>`. Personally I prefer having the `RUN` that needs line next to the actual line itself and then using `%(line+1)`. But I seem to be the only one, so up to you. It would let you move the build to the top of the file though 🤷

hamishknight replied:
Sure, updated to use `split-file`, and also updated a couple other variants of this test to do so. Unfortunately it doesn't seem like `%(line+1)` works for anything other than the first split file, I get the error `wrong pos format, line too large`, is that expected? In any case, I've moved the other `RUN` lines to the top of the file, as those won't affect the line position for the split file anyways.

bnbarham replied:
You can add `--leading-lines` to handle those cases. It'll add padding newlines to the start of each file so that their lines reflect the original source file.

hamishknight replied:
Cool!

bnbarham commented on test/SourceKit/CodeComplete/complete_docbrief_package.swift:
```diff
@@ -0,0 +1,50 @@
+// CHECK: {
+// CHECK:   key.results: [
+// CHECK-NEXT:     {
+// CHECK-NEXT:       key.kind: source.lang.swift.decl.function.method.instance,
```
I'd prefer only checking the things we're actually testing here, just name + doc is probably enough (same for the other tests).

hamishknight replied:
Updated to drop some of the expectations, I would like to keep the expectations between the name and doc comment though, as that lets us continue to use `CHECK-NEXT`, instead of needing to put in a `CHECK-NOT: }`, which reads weirdly IMO.

hamishknight commented:
@swift-ci please test

hamishknight commented on lib/AST/RawComment.cpp:
```diff
@@ -130,60 +131,43 @@ static RawComment toRawComment(ASTContext &Context, CharSourceRange Range) {
-
-  // Check the cache in ASTContext.
-  auto &Context = getASTContext();
-  if (Optional<std::pair<RawComment, bool>> RC = Context.getRawComment(this)) {
-    auto P = RC.value();
-    if (!SerializedOK || P.second)
-      return P.first;
-  }
```
Heh, just noticed this predicate was wrong anyway, and allowed us to retrieve a serialized entry from the cache even if `!SerializedOK`.

hamishknight commented:
@swift-ci please test

hamishknight commented on lib/Serialization/SerializeDoc.cpp:
```diff
@@ -309,53 +310,15 @@ static void writeGroupNames(const comment_block::GroupNamesLayout &GroupNames,
-  }
-
-  // Skip SPI decls, unless we're generating a symbol graph with SPI information.
-  if (D->isSPI() && !D->getASTContext().SymbolGraphOpts.IncludeSPISymbols)
-    return false;
```
@QuietMisdreavus In factoring out `getDocCommentSerializationTargetFor`, I removed the check for `SymbolGraphOpts.IncludeSPISymbols` here so that the check is consistent for both serialization and deserialization. I don't believe this should be an issue though, as we now are able to serialize SPI decl comments in the swiftsourceinfo. Any concerns with this change?

QuietMisdreavus replied:
Is there any way to remove SPI information from `.swiftdoc` files after this change? I'm wondering about potential leaks of implementation details if someone had a doc comment that wasn't meant for the light of day landing in a file like this.

bnbarham replied:
To be clear, with this change there *isn't* any SPI information in swiftdoc. Where as previously if `SymbolGraphOpts.IncludeSPISymbols` was `true` it could have been. Now it is *only* serialized in swiftsourceinfo.

But in any case, isn't that up to whether symbol graph outputs the symbol or not? If it does then the symbol is leaked, irrespective of the documentation being leaked.

QuietMisdreavus replied:
Ah, i misunderstood then. The symbol graph only has SPI information if you specifically ask for it, but i wasn't sure if it was going to inadvertently show up in other places where it wasn't desired. This makes sense, thanks!

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

bnbarham commented on lib/Serialization/SerializeDoc.cpp:
```diff
@@ -394,7 +357,9 @@ static void writeDeclCommentTable(
         return false;
 
-      // Skip the decl if it does not have a comment.
+      // Skip the decl if it does not have a comment. Note this means
+      // we'll only serialize "direct" brief comments, but that's okay
+      // because clients can compute the semantic brief comment themselves.
```
Seems weird that we output the brief comment to me at all. @nkcsgexi do you remember why this is the case?

bnbarham reviewed (approved):
LGTM, thanks Hamish. Might be good to get a review from @nkcsgexi as well.

nkcsgexi commented:
oh, this is a great idea to include doc comment in the source info file!

hamishknight committed:
Requestify raw and brief comment computation

Turn these ASTContext cached computations into
request evaluator cached computations.

hamishknight committed:
Drop `SerializedOK` parameter from `getRawComment`

It doesn't seem like there's any client that's
actually taking advantage of setting it to `false`,
and its default value of `false` is more likely
than not going to cause clients to accidentally
miss comments that they may want. In fact, this
was exactly the case for code completion's brief
field. Finally, the parameter wasn't even
consistently applied, as we would attempt to
deserialize swiftdoc comments even if it were
`false`.

hamishknight committed:
Refactor a couple of tests

Match the format of the spi and package variants of
these tests.

hamishknight committed:
Serialize `package` and SPI doc comments in swiftsourceinfo

Previously we were using the same set of conditions
for serializing as for swiftdoc, so excluded them.
However it's reasonable to have them in the
swiftsourceinfo.

hamishknight committed:
Adjust BriefCommentRequest to only query swiftdoc if we have it

If we have both loaded a swiftdoc, and the decl we
have should have had its doc comment serialized into
it, we can check it without needing to fall back
to the swiftsourceinfo.

This requires a couple of refactorings:

- Factoring out the `shouldIncludeDecl` logic
into `getDocCommentSerializationTargetFor` for
determining whether a doc comment should end up
in the swiftdoc or not.
- Factoring out `CommentProviderFinder` for searching
for the doc providing comment decl for brief
comments, in order to allow us to avoid querying
the raw comment when searching for it. This has the
added bonus of meaning we no longer need to fall
back to parsing the raw comment for the brief
comment if the comment is provided by another decl
in the swiftdoc.

This diff is best viewed without whitespace.

hamishknight committed:
Rename `getBriefComment` -> `getSemanticBriefComment`

Make it clear that we will walk the comment providing
decls if needed.

hamishknight committed:
Update to use `--leading-lines` for `split-file`

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Linux

hamishknight commented:
Happy to take care of further feedback in a follow-up
